scheduler/scheduler.go

```diff
@@ -118,7 +118,7 @@ type Scheduler interface {
 	Runner
 
 	// Submit submits an app, creating it or updating it as necessary.
-	Submit(context.Context, *App) error
+	Submit(context.Context, *App, EventChan) error
```

Comment:
The more I think about it, the more I think an interface here would be better, and prevent easy mistakes with channels from happening.

My first thought, would be an interface like this:

```go
type Events interface {
        // Publish an event to the event stream.
        Publish(Event)

        // Signals that no more events will be published.
        Done(error)
}
```

We can then implement a channel based implementation of this, so that publishers don't have to worry about blocking:

```go
import "sync"

type events struct {
	sync.Mutex
	done bool
	err  error
	ch   chan Event
}

func newEvents() *events {
	return &events{ch: make(chan Event, 100)}
}

func (e *events) Publish(event Event) {
	e.Lock()
	defer e.Unlock()

	if done {
		// Or just warn.
		panic("Publish called on finalized event stream")
	}

	e.publish(event)
}

func (e *events) publish(event Event) {
	select {
	case e.ch <- event:
	default:
		// Drop
	}
}

func (e *events) Subscribe() <-chan Event {
	return e.ch
}

func (e *events) Done(err error) {
	e.Lock()
	defer e.Unlock()

	e.done = true
	e.err = err
}

func (e *events) Err() error {
	return e.err
}
```

So then, submitting and consuming the event stream would look like this:

```go
events := scheduler.NewEvents()
s.Submit(ctx, app, events)

for event := range events {
        // Write jsonmessage to stream.
}

if err := events.Err(); err != nil {
        // Write error to stream.
}
```